src/os/ObjectStore.h

```diff
@@ -1718,23 +1719,78 @@ class ObjectStore {
    * @param op_flags is CEPH_OSD_OP_FLAG_*
    * @returns number of bytes read on success, or negative error code on failure.
    */
-   virtual int read(
-    const coll_t& cid,
+  virtual int read(
+   const coll_t& cid,
+   const ghobject_t& oid,
+   uint64_t offset,
+   size_t len,
+   bufferlist& bl,
+   uint32_t op_flags = 0) = 0;
+  virtual int read(
+    CollectionHandle &c,
     const ghobject_t& oid,
     uint64_t offset,
     size_t len,
     bufferlist& bl,
-    uint32_t op_flags = 0) = 0;
-   virtual int read(
-     CollectionHandle &c,
-     const ghobject_t& oid,
-     uint64_t offset,
-     size_t len,
-     bufferlist& bl,
-     uint32_t op_flags = 0) {
-     return read(c->get_cid(), oid, offset, len, bl, op_flags);
-   }
+    uint32_t op_flags = 0) {
+    return read(c->get_cid(), oid, offset, len, bl, op_flags);
+  }
+
+  virtual bool has_async_read() const {
+    return false;
+  }
+
+  struct async_read_params_t {
+    uint64_t offset = 0;
+    uint64_t length = 0;
+    ceph::bufferlist* outbl = nullptr;
+    Context* on_complete = nullptr;
+    uint32_t flags = 0;
+
+    async_read_params_t(
+      const uint64_t offset,
+      const uint64_t length,
+      ceph::bufferlist* const outbl,
+      Context* const on_complete,
+      const uint32_t flags)
+    : offset(offset),
+      length(length),
+      outbl(outbl),
+      on_complete(on_complete),
+      flags(flags) {
+    }
+  };
 
+  /**
+   * async_read -- asynchronously read a byte range of data from an object
+   *
+   * Note: if reading from an offset past the end of the object, we
+   * return 0 (not, say, -EINVAL).
+   *
+   * @param cid collection for object
+   * @param oid oid of object
+   * @param offset location offset of first byte to be read
```

Comment:
FIXME: change the description to reflect the `async_read_params_t`-related rework.